### PR TITLE
Implement kafka-file fault tolerant parser.

### LINKED ISSUE #142: Implement kafka-file fault tolerant parser.
To support properly completion, validation inside kafka file (see https://github.com/jlandersen/vscode-kafka/issues/92) we need to implement a kafka-file fault tolerant parser (AST kind).

### PR DESCRIPTION
**Pull request by angelozerr**
Implement kafka-file fault tolerant parser.

Fixes #142

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This draft PR move the parse stuff in a kafka file parser which build an AST KafkaFileDocument only when text content changed (and not each time code lens is refreshed).

I must clean my codeagain and I need to write tests.

@fbricon if you wish you can check this PR doesn't break the codelens feature of kafka file. I need to clean my code again.

File: src/kafka-file/kafkaFileClient.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        return runSafeAsync(async () => {
+            const kafkaFileDocument = this.kafkaFileDocuments.get(document);
+            return this.languageService.getCodeLenses(document, kafkaFileDocument);
+        }, [], `Error while computing code lenses for ${document.uri}`, token);
+    }
+
+    dispose() {
+        this.disposables.forEach(d => d.dispose());
+    }
+}
```
Comment (fbricon) [inline src/kafka-file/kafkaFileClient.ts:48]: I'd move to its own file

Comment (angelozerr) [reply src/kafka-file/kafkaFileClient.ts]: This class is just a wrapper which consumes language service, so I would like to keep it in this client file. I restructured the file, please tell me if it's better.

File: src/kafka-file/languageservice/parser/kafkaFileParser.ts [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+}
+function isEndBlock(lineText: string, blockType: BlockType): boolean {
+    if (blockType === BlockType.consumer) {
+        return isSeparator(lineText) || getBlockType(lineText) !== undefined;
+    }
+    return isSeparator(lineText);
+}
+
+function isSeparator(lineText: string): boolean {
+    return lineText === "###";
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply src/kafka-file/languageservice/parser/kafkaFileParser.ts]: fixed

File: src/test/suite/kafka-file/languageservice/kafkaAssert.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+export function getSimpleLanguageService() {
+    return languageService;
+}
+export function getDocument(content: string) {
+    return workspace.openTextDocument({
+        language: 'kafka',
+        content
+    });
+}
+export function p(startLine: number, startCharacter: number): Position {
```
Comment (fbricon) [inline src/test/suite/kafka-file/languageservice/kafkaAssert.ts:66]: I'd prefer a more explicit name: position

File: src/test/suite/kafka-file/languageservice/kafkaAssert.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    return new Position(startLine, startCharacter);
+}
+
+// Code Lens assert
+
+export function cl(start: Position, end : Position, command: Command) : CodeLens {
```
Comment (fbricon) [inline src/test/suite/kafka-file/languageservice/kafkaAssert.ts:72]: function codeLens

File: src/test/suite/kafka-file/languageservice/kafkaAssert.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+// Kafka parser assert
+export interface ExpectedBlockResult {
+    type: BlockType;
+    start: Position;
+    end: Position;
+    properties?: Array<Property>
+}
+
+export function b(type: BlockType, start: Position, end: Position, properties?: Array<Property>): ExpectedBlockResult {
```
Comment (fbricon) [inline src/test/suite/kafka-file/languageservice/kafkaAssert.ts:91]: function block

**Review by fbricon — CHANGES_REQUESTED** [review]
I understand the appeal of short method names, for writing, but it makes the code harder to read/grok. And we spend more time reading than writing, so I'd rather sacrifice the latter a little bit

File: src/test/suite/kafka-file/languageservice/kafkaAssert.ts (hunk 3 above)
Comment (angelozerr) [reply src/test/suite/kafka-file/languageservice/kafkaAssert.ts]: fixed

File: src/test/suite/kafka-file/languageservice/kafkaAssert.ts (hunk 4 above)
Comment (angelozerr) [reply src/test/suite/kafka-file/languageservice/kafkaAssert.ts]: fixed

File: src/test/suite/kafka-file/languageservice/kafkaAssert.ts (hunk 5 above)
Comment (angelozerr) [reply src/test/suite/kafka-file/languageservice/kafkaAssert.ts]: fixed

**Comment by fbricon** [comment]
Thanks @angelozerr

**Pull request merged**